### DISCUSSION TITLE
applying  -contrast-stretch to a smaller polygon part of an image

### DISCUSSION THREAD

**Discussion by thisisnottaken** [ImageMagick/ImageMagick#4668] [Digital Image Processing]
I have an image in which there is a specific area in which I want more contrast .
How can I create a mask (polygon shape whose vertices coordinates I can give as input) and apply contrast stretch to only that part of image ?

Thanks.

**Comment by fmw42** [comment]
Use -draw "polygon .... " to make your mask. Then apply the contrast stretch to the full image. Then use the mask the composite the original and contrast stretched images together. See 

https://legacy.imagemagick.org/Usage/compose/#compose
https://legacy.imagemagick.org/Usage/draw/#primitives

For further help, always provide your IM version and platform/OS as syntax may vary.

**Comment by snibgo** [comment]
As fmw42 says. However, "-contrast-stretch" does two things:

1. It uses statistics from the image to decide how much to stretch.

2. It applies the stretch.

So a "-contrast-stretch" of part of the image will give a different result, for that part, than a "-contrast-stretch" of the entire image. If that matters to you, please give more details, eg sample input and polygon, as well as the IM version and platform.

**Comment by thisisnottaken** [comment]
![rotated-dl10sz3586](https://user-images.githubusercontent.com/19471922/148411507-2ddd00a5-2d08-4112-912f-35636b0ef56e.jpg)

In this image polygon will be 1619,2550 1855,2516 1874,2661 1646,2697
Assuming top left corner of image is 0,0

I am using ubuntu 64 bit OS with
$ convert --version
Version: ImageMagick 6.8.9-9 Q16 x86_64 2020-12-04 http://www.imagemagick.org
Copyright: Copyright (C) 1999-2014 ImageMagick Studio LLC
Features: DPC Modules OpenMP
Delegates: bzlib cairo djvu fftw fontconfig freetype jbig jng jpeg lcms lqr ltdl lzma openexr pangocairo png rsvg tiff wmf x xml zlib

**Comment by snibgo** [comment]
Your IM is very old. I suggest you upgrade, preferably to v7.

If you want to use the statistics from just that polygon, "-contrast-stetch" won't help because it doesn't use a read-mask. Instead, you have to work from first principles.

Suppose you want to stretch so the darkest 10% become black, and the lightest 10% become white. Invert the cumulative histogram, and the pixel at 0.1\*width gives the bottom level, and the pixel at 0.9\*width gives the top level, for a "-level" operation.

Here is a Windows BAT script that uses IM v7, with my process modules:
```
magick ^
  motcycl.jpg ^
  +write mpr:INP ^
  ( +clone ^
    -fill Black -colorize 100 ^
    -fill White -draw "polygon 1619,2550 1855,2516 1874,2661 1646,2697" ^
  ) ^
  -alpha off ^
  -compose CopyOpacity -composite ^
  -process 'mkhisto regardalpha cumul norm' ^
  -process invclut ^
  ( +clone -crop 1x1+%%[fx:w*0.1]+0 +repage +write txt: -set option:LO %%[fx:mean*QuantumRange] +delete ) ^
  ( +clone -crop 1x1+%%[fx:w*0.9]+0 +repage +write txt: -set option:HI %%[fx:mean*QuantumRange] +delete ) ^
  +delete ^
  mpr:INP ^
  -level %%[LO]x%%[HI] ^
  x.tiff
```

**Reply by CelestialStreamer** [reply]
I'm using ImageMagick 7.1.0-25 Q16-HDRI x64 on Windows and I got an error
```
magick.exe: unable to open module file 'C:\Program Files\ImageMagick-7.1.1-Q16-HDRI\modules\filters\FILTER_mkhisto_.dll': No such file or directory @ warning/module.c/GetMagickModulePath/677.
magick.exe: unable to load module 'FILTER_mkhisto_.dll': C:\Program Files\ImageMagick-7.1.1-Q16-HDRI\modules\filters\FILTER_mkhisto_.dll @ error/module.c/InvokeDynamicImageFilter/990.
```

Indeed in my installation folder `modules/filters` there is only `FILTER_analyze_.dll` and not `FILTER_mkhisto_.dll`.

Could you either tell me how to install those filters, or how to rewrite your command so as not to depend on `-process 'whatever it is you're doing here'`?

**Comment by fmw42** [comment]
I do not see the point of using -contrast-stretch on the whole image from the statistics of a subsection. -contrast-stretch computes statistics not by gray level but by percent from the ends of the histogram. The clipping for one region may not be relevant to any other region or the whole image, because the statistics for the whole image is different from that of a subsection.

**Comment by GeeMack** [comment]
There are several ways to modify the contrast with ImageMagick, so depending on how precise your requirements are, one way may be better than another. Here's a command that works in IMv6.8.9-9 that isolates a region, modifies the contrast to enhance the numbers on the license plate, and saves the result...

    convert image.png \
       \( \
          +clone -evaluate set 100% \
          -draw "polygon 1619,2550 1855,2516 1874,2661 1646,2697" \
          -write mpr:mask +delete \
       \) \
       -mask mpr:mask -brightness-contrast 0,-50 +mask -chop 50% result.png

![result](https://user-images.githubusercontent.com/65475279/148449882-f2a0c226-0953-46c2-90c5-a93a43671beb.png)

That makes a copy of the image inside the parentheses, colors it white with `-evaluate set 100%`, and draws a black polygon on it. After the parentheses it uses that as a write mask so the contrast operation only affects the selected area. Some adjustment to the arguments to the `-brightness-contrast` operation may get you near what you're trying to achieve.

**Reply by thisisnottaken** [reply]
i tried your command . i think you have posted result.png in which there is no improvement from original image . currently argument to -brightness-contrast  is 0,-50 . what range of values should I try ?

**Reply by CelestialStreamer** [reply]
I cannot use `-mask` and `+mask`:
```
magick.exe: option deprecated, unable to execute `-mask' at CLI arg 15 @ error/operation.c/CLIOption/5418.
```

Indeed, the [documentation](https://imagemagick.org/script/command-line-options.php) only mentions `-clip` and `-clip-mask`, either of which I still can't figure out how to use.

Can you modify your command to work with >=7?

**Comment by snibgo** [comment]
> I do not see the point of using -contrast-stretch on the whole image from the statistics of a subsection.

I suppose the task is to make the licence plate more legible by increasing the contrast. For that, we need the statistics from just the licence plate. That gives us the numbers for a "-level" command, which we might use on the entire image, or just a crop, or the entire image which is then composited over the original with a mask so only the licence plate is changed, or whatever.

**Reply by thisisnottaken** [reply]
yes snibgo . here i wanted to increase legibility of number plate without modifying other parts of image .
Another case is when I have a image containing text and photograph and my aim would be to improve text readability without touching the photograph . Assume that photograph is not in the middle so that I can define a polygon that steers clear of photograph .

**Comment by snibgo** [comment]
```
magick ^
  motcycl.jpg ^
  +write mpr:INP ^
  ( +clone ^
    -fill Black -colorize 100 ^
    -fill White -draw "polygon 1619,2550 1855,2516 1874,2661 1646,2697" ^
    -alpha off ^
    +write mpr:MASK ^
  ) ^
  -alpha off ^
  -compose CopyOpacity -composite ^
  -process 'mkhisto regardalpha cumul norm' ^
  -process invclut ^
  ( +clone -crop 1x1+%%[fx:w*0.05]+0 +repage +write txt: -set option:LO %%[fx:mean*QuantumRange] +delete ) ^
  ( +clone -crop 1x1+%%[fx:w*0.95]+0 +repage +write txt: -set option:HI %%[fx:mean*QuantumRange] +delete ) ^
  +delete ^
  mpr:INP ^
  -level %%[LO]x%%[HI] ^
  mpr:INP ^
  +swap ^
  mpr:MASK ^
  -compose Over -composite ^
  x.jpg
```
![x](https://user-images.githubusercontent.com/33812368/148492289-4075d56e-70aa-428f-bb23-64c6ad265eae.jpg)

**Reply by thisisnottaken** [reply]
i understood that contrast-stretch does not work on mask . The command above did not work for my imagemagick on ubuntu . Also, is there a reason for yellow color in number plate ?

**Comment by snibgo** [comment]
My commands above use my process modules "mkhisto" and "invclut". The commands won't work unless IM is built with those process modules. Instructions for including the process modules are at [Process modules](http://im.snibgo.com/customim.htm).

Some of the pixels in the licence plate are yellow, but low saturation. Increasing contrast usually also increases saturation.

Your input is JPG, which is only 8 bits per channel. Extreme changes to contrast will exaggerate quantization problems, introducing banding, and shifts to saturation and hue.

**Reply by thisisnottaken** [reply]
ok . I want number plate to be legible and also look real . How can I sharpen the text in number plate ?
